automod/engine/context.go

```diff
@@ -0,0 +1,184 @@
+	AccountContext
+
+	RecordOp RecordOp
+	// TODO: could consider adding commit-level metadata here. probably nullable if so, commit-level metadata isn't always available. might be best to do a separate event/context type for that
+}
+
+var (
+	CreateOp = "create"
+	UpdateOp = "update"
+	DeleteOp = "delete"
+)
+
+// Immutable
+type RecordOp struct {
+	// Indicates type of record mutation: create, update, or delete.
+	// The term "action" is copied from com.atproto.sync.subscribeRepos#repoOp
+	Action     string
+	DID        syntax.DID
+	Collection syntax.NSID
+	RecordKey  syntax.RecordKey
+	CID        *syntax.CID
+	// NOTE: usually a *pointer*, not the value itself
+	Value any
+}
+
+// Checks that op has expected fields, based on the action type
+func (op *RecordOp) Validate() error {
+	switch op.Action {
+	case CreateOp, UpdateOp:
+		if op.Value == nil || op.CID == nil {
+			return fmt.Errorf("expected record create/update op to contain both value and CID")
+		}
+	case DeleteOp:
+		if op.Value != nil || op.CID != nil {
+			return fmt.Errorf("expected record delete op to be empty")
+		}
+	default:
+		return fmt.Errorf("unexpected record op action: %s", op.Action)
+	}
+	return nil
+}
+
+func (op *RecordOp) ATURI() syntax.ATURI {
+	return syntax.ATURI(fmt.Sprintf("at://%s/%s/%s", op.DID, op.Collection, op.RecordKey))
+}
+
+// TODO: in the future *may* have an IdentityContext with an IdentityOp sub-field
+
+// Access to engine's identity directory (without access to other engine fields)
+func (c *BaseContext) Directory() identity.Directory {
+	return c.engine.Directory
+}
+
+// request external state via engine (indirect)
+func (c *BaseContext) GetCount(name, val, period string) int {
+	out, err := c.engine.Counters.GetCount(c.Ctx, name, val, period)
+	if err != nil {
+		if nil == c.Err {
+			c.Err = err
+		}
+		return 0
+	}
+	return out
+}
+
+func (c *BaseContext) GetCountDistinct(name, bucket, period string) int {
+	out, err := c.engine.Counters.GetCountDistinct(c.Ctx, name, bucket, period)
+	if err != nil {
+		if nil == c.Err {
+			c.Err = err
+		}
+		return 0
+	}
+	return out
+}
+
+func (c *BaseContext) InSet(name, val string) bool {
+	out, err := c.engine.Sets.InSet(c.Ctx, name, val)
+	if err != nil {
+		if nil == c.Err {
+			c.Err = err
+		}
+		return false
+	}
+	return out
+}
+
+func NewAccountContext(ctx context.Context, eng *Engine, meta AccountMeta) AccountContext {
+	return AccountContext{
+		BaseContext: BaseContext{
+			Ctx:     ctx,
+			Err:     nil,
+			Logger:  eng.Logger.With("did", meta.Identity.DID),
+			engine:  eng,
+			effects: Effects{},
+		},
+		Account: meta,
+	}
+}
+
+func NewRecordContext(ctx context.Context, eng *Engine, meta AccountMeta, op RecordOp) RecordContext {
+	ac := NewAccountContext(ctx, eng, meta)
+	ac.BaseContext.Logger = ac.BaseContext.Logger.With("collection", op.Collection, "rkey", op.RecordKey)
+	return RecordContext{
+		AccountContext: ac,
+		RecordOp:       op,
+	}
+}
+
+// update effects (indirect)
```

Comment:
I'd vote weakly against these proxy methods and prefer exposing the Effects struct (either as `Effects` a field, or `Effects()` a getter; either works).   Mainly because I'm a fan of the guidance/focus that autocomplete can give if I start with `ctx.Effects.` whenever I'm looking for a mutation method and I can't remember the name.

Field or method doesn't feel like it matters much.  The main thing exposing it as a field directly changes is it makes it zeroable by out-of-package code.  If we were in a language with a stronger typesystem, I'd say that's worse, but in golang, I'd say it's par.

Reply:
One thing I don't like about direct Effects access is splitting things like `GetCount` and `Increment` across two types. Feels like that splits documentation/awareness, though it does (maybe) communicate what methods have effects and which don't.